File: internal/parser/scanner.go
// scanner.go: kvlang 词法分析。
// Scan(src string) → []Token  支持整文件多行扫描，末尾附 EOF 哨兵。
package parser

import (
	"fmt"
	"strings"
)

// Pos 携带 Token 在源码中的起始位置。
type Pos struct {
	Line int // 1-based
	Col  int // 1-based
}

// Diagnostic 表示一条解析错误，携带位置信息。
type Diagnostic struct {
	Pos     Pos
	Message string
}

func (d Diagnostic) String() string {
	return fmt.Sprintf("%d:%d: %s", d.Pos.Line, d.Pos.Col, d.Message)
}

// Kind 标记 Token 类型。
type Kind int

const (
	Ident    Kind = iota // 标识符、符号算子、未引用数字/路径段
	Literal              // 字面量（引号已去除）
	Arrow                // -> 或 <-
	LParen               // (
	RParen               // )
	Comma                // ,
	LBrace               // {
	RBrace               // }
	Colon                // :
	Return               // return
	If                   // if
	Else                 // else
	For                  // for
	While                // while
	Break                // break
	Continue             // continue
	Newline              // 换行（语句分隔符，连续换行折叠为一个）
	Comment              // # 行注释（含 # 字符）
	EOF                  // 文件结束哨兵
)

// String 返回 Kind 的可读名称。
func (k Kind) String() string {
	names := [...]string{
		"IDENT", "LITERAL", "ARROW",
		"LPAREN", "RPAREN", "COMMA", "LBRACE", "RBRACE", "COLON",
		"RETURN", "IF", "ELSE", "FOR", "WHILE",
		"BREAK", "CONTINUE",
		"NEWLINE", "COMMENT", "EOF",
	}
	if int(k) < len(names) {
		return names[k]
	}
	return "UNKNOWN"
}

// Token 表示一个词法单元，携带源码起始位置。
type Token struct {
	Kind  Kind
	Value string
	Pos   Pos
}

// String 返回 Token 的调试表示。
func (t Token) String() string {
	return fmt.Sprintf("%s(%q)@%d:%d", t.Kind, t.Value, t.Pos.Line, t.Pos.Col)
}

// 语言关键字字符串常量（供 keywords 表和注释使用）。
const kwReturn   = "return"
const kwIf       = "if"
const kwElse     = "else"
const kwFor      = "for"
const kwWhile    = "while"
const kwBreak    = "break"
const kwContinue = "continue"

// keywords 将语言关键字映射到对应 Token Kind。
var keywords = map[string]Kind{
	kwReturn:   Return,
	kwIf:       If,
	kwElse:     Else,
	kwFor:      For,
	kwWhile:    While,
	kwBreak:    Break,
	kwContinue: Continue,
}

// singleCharToken 将单字符标点映射到对应 Kind。
var singleCharToken = map[byte]Kind{
	'(': LParen, ')': RParen, ',': Comma,
	'{': LBrace, '}': RBrace, ':': Colon,
}

// scanQuoted 从 src[i]（引号字符）开始，返回引号内容和结束后的下一位置。
func scanQuoted(src string, i int, quote byte) (string, int) {
	end := strings.IndexByte(src[i+1:], quote)
	if end >= 0 {
		return src[i+1 : i+1+end], i + end + 2
	}
	return src[i+1:], len(src)
}

// Scan 将整个源字符串（可含换行）扫描为平坦 Token 流，末尾附 EOF 哨兵。
//
// 每个 Token 携带源码起始位置 Pos{Line, Col}（均 1-based）。
// # 注释产生 Comment Token，格式化工具可通过 Comments 字段保留注释（S6）。
// 连续换行折叠为一个 Newline Token。
func Scan(src string) []Token {
	var tokens []Token
	i := 0
	line := 1
	lineStart := 0  // 当前行起始字节偏移
	prevNewline := true

	// pos 返回当前字节偏移 i 对应的源码位置。
	pos := func() Pos { return Pos{Line: line, Col: i - lineStart + 1} }

	for i < len(src) {
		c := src[i]

		// 换行：折叠连续换行
		if c == '\n' {
			if !prevNewline && len(tokens) > 0 {
				tokens = append(tokens, Token{Kind: Newline, Value: "\n", Pos: pos()})
				prevNewline = true
			}
			i++
			line++
			lineStart = i
			continue
		}
		if c == '\r' {
			i++
			continue
		}

		// 空白（非换行）
		if c == ' ' || c == '\t' {
			i++
			continue
		}

		// ';' — 显式语句分隔符，折叠规则与 '\n' 相同
		if c == ';' {
			if !prevNewline && len(tokens) > 0 {
				tokens = append(tokens, Token{Kind: Newline, Value: ";", Pos: pos()})
				prevNewline = true
			}
			i++
			continue
		}

		// # 行注释 → Comment Token（S6：保留注释）
		if c == '#' {
			p := pos()
			start := i
			for i < len(src) && src[i] != '\n' {
				i++
			}
			tokens = append(tokens, Token{Kind: Comment, Value: src[start:i], Pos: p})
			prevNewline = false // Comment 是实质 token，后续 \n 会触发 Newline
			continue
		}

		prevNewline = false
		p := pos() // 记录当前 token 起始位置

		// 引号字符串 → Literal（引号去除）
		if c == '\'' || c == '"' {
			val, next := scanQuoted(src, i, c)
			tokens = append(tokens, Token{Kind: Literal, Value: val, Pos: p})
			i = next
			continue
		}

		// 左箭头 <-
		if c == '<' && i+1 < len(src) && src[i+1] == '-' {
			tokens = append(tokens, Token{Kind: Arrow, Value: "<-", Pos: p})
			i += 2
			continue
		}

		// 右箭头 ->
		if c == '-' && i+1 < len(src) && src[i+1] == '>' {
			tokens = append(tokens, Token{Kind: Arrow, Value: "->", Pos: p})
			i += 2
			continue
		}

		// 双字符算子（在单字符之前匹配）
		if i+1 < len(src) {
			switch src[i : i+2] {
			case "==", "!=", "<=", ">=", "&&", "||", "<<", ">>":
				tokens = append(tokens, Token{Kind: Ident, Value: src[i : i+2], Pos: p})
				i += 2
				continue
			}
		}

		// 单字符标点 — 查表
		if k, ok := singleCharToken[c]; ok {
			tokens = append(tokens, Token{Kind: k, Value: string(c), Pos: p})
			i++
			continue
		}

		// 单字符符号算子
		switch c {
		case '+', '*', '%', '!', '<', '>', '&', '|', '^':
			tokens = append(tokens, Token{Kind: Ident, Value: string(c), Pos: p})
			i++
			continue
		}

		// '-' 单独处理（已排除 -> 的情况）
		if c == '-' {
			tokens = append(tokens, Token{Kind: Ident, Value: "-", Pos: p})
			i++
			continue
		}

		// 数字字面量
		if c >= '0' && c <= '9' {
			start := i
			for i < len(src) && (src[i] >= '0' && src[i] <= '9' || src[i] == '.' || src[i] == 'e' || src[i] == 'E') {
				i++
			}
			tokens = append(tokens, Token{Kind: Literal, Value: src[start:i], Pos: p})
			continue
		}

		// 路径字面量 ./foo
		if c == '.' && i+1 < len(src) && src[i+1] == '/' {
			start := i
			i += 2
			for i < len(src) && !isTokenDelim(src[i]) {
				i++
			}
			tokens = append(tokens, Token{Kind: Literal, Value: src[start:i], Pos: p})
			continue
		}

		// '/' — 绝对路径字面量 或 除法算子
		if c == '/' {
			if i+1 < len(src) && isAbsPathStart(src[i+1]) {
				start := i
				i++
				for i < len(src) && !isTokenDelim(src[i]) {
					i++
				}
				tokens = append(tokens, Token{Kind: Literal, Value: src[start:i], Pos: p})
			} else {
				tokens = append(tokens, Token{Kind: Ident, Value: "/", Pos: p})
				i++
			}
			continue
		}

		// 关键字 / 标识符（含点号：tensor.new 等）
		start := i
		for i < len(src) && !isTokenDelim(src[i]) {
			i++
		}
		if i == start {
			i++ // 跳过无法识别的字符，防止无限循环
			continue
		}
		word := src[start:i]
		if k, ok := keywords[word]; ok {
			tokens = append(tokens, Token{Kind: k, Value: word, Pos: p})
		} else {
			tokens = append(tokens, Token{Kind: Ident, Value: word, Pos: p})
		}
	}

	tokens = append(tokens, Token{Kind: EOF, Value: "", Pos: pos()})
	return tokens
}

func isTokenDelim(c byte) bool {
	switch c {
	case ' ', '\t', '\n', '\r', ';',
		',', ')', '(', '{', '}',
		'+', '-', '*', '%', '!', '=', '<', '>', '&', '|', '^',
		':':
		return true
	}
	return false
}

func isAbsPathStart(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
		(c >= '0' && c <= '9') || c == '_'
}

File: layout/src/scanner.rs
//! 词法分析（对齐 parser/scanner.go）。Scan(src) → Vec<Token>，末尾附 EOF 哨兵。

use std::fmt;

use super::symbol;

#[derive(Clone, Copy)]
pub struct Pos {
    pub line: i32, // 1-based
    pub col: i32,  // 1-based
}

#[derive(Clone)]
pub struct Diagnostic {
    pub pos: Pos,
    pub message: String,
    pub warn: bool, // true = 警告
    pub info: bool, // true = 提示（优先级高于 warn）
    pub source: String,
    pub src_file: String,
    pub src_name: String,
}

impl Diagnostic {
    pub fn string(&self) -> String {
        let kind = if self.info {
            "info"
        } else if self.warn {
            "warn"
        } else {
            "error"
        };
        let src = if self.src_name.is_empty() {
            self.src_file.as_str()
        } else {
            self.src_name.as_str()
        };
        if !src.is_empty() {
            format!(
                "{kind}: {src}:{}:{}: {}",
                self.pos.line, self.pos.col, self.message
            )
        } else {
            format!(
                "{kind}: {}:{}: {}",
                self.pos.line, self.pos.col, self.message
            )
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    Ident,
    Literal,
    Arrow,
    LParen,
    RParen,
    Comma,
    LBrace,
    RBrace,
    LBrack,
    RBrack,
    Dot,
    Colon,
    Return,
    If,
    Else,
    For,
    While,
    Break,
    Continue,
    Newline,
    Comment,
    EOF,
}

impl Kind {
    pub fn to_str(self) -> &'static str {
        match self {
            Kind::Ident => "IDENT",
            Kind::Literal => "LITERAL",
            Kind::Arrow => "ARROW",
            Kind::LParen => "LPAREN",
            Kind::RParen => "RPAREN",
            Kind::Comma => "COMMA",
            Kind::LBrace => "LBRACE",
            Kind::RBrace => "RBRACE",
            Kind::LBrack => "LBRACK",
            Kind::RBrack => "RBRACK",
            Kind::Dot => "DOT",
            Kind::Colon => "COLON",
            Kind::Return => "RETURN",
            Kind::If => "IF",
            Kind::Else => "ELSE",
            Kind::For => "FOR",
            Kind::While => "WHILE",
            Kind::Break => "BREAK",
            Kind::Continue => "CONTINUE",
            Kind::Newline => "NEWLINE",
            Kind::Comment => "COMMENT",
            Kind::EOF => "EOF",
        }
    }
}

impl fmt::Display for Kind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_str())
    }
}

#[derive(Clone)]
pub struct Token {
    pub kind: Kind,
    pub value: String,
    pub pos: Pos,
    pub quote: u8, // 0=无, '"'=转义串 "...", 'r'=原始串 r#"..."#
}

impl Token {
    pub fn eof(pos: Pos) -> Token {
        Token {
            kind: Kind::EOF,
            value: String::new(),
            pos,
            quote: 0,
        }
    }
}

fn keyword(s: &str) -> Option<Kind> {
    Some(match s {
        "return" => Kind::Return,
        "if" => Kind::If,
        "else" => Kind::Else,
        "for" => Kind::For,
        "while" => Kind::While,
        "break" => Kind::Break,
        "continue" => Kind::Continue,
        _ => return None,
    })
}

fn single_char_token(c: u8) -> Option<Kind> {
    Some(match c {
        b'(' => Kind::LParen,
        b')' => Kind::RParen,
        b',' => Kind::Comma,
        b'{' => Kind::LBrace,
        b'}' => Kind::RBrace,
        b'[' => Kind::LBrack,
        b']' => Kind::RBrack,
        b':' => Kind::Colon,
        _ => return None,
    })
}

fn escaped_byte(c: u8) -> u8 {
    match c {
        b'n' => b'\n',
        b't' => b'\t',
        b'r' => b'\r',
        b'0' => 0,
        other => other,
    }
}

fn scan_quoted(src: &[u8], mut i: usize, quote: u8) -> (String, usize) {
    i += 1;
    let mut b: Vec<u8> = Vec::new();
    while i < src.len() {
        let c = src[i];
        if c == b'\\' {
            i += 1;
            if i < src.len() {
                b.push(escaped_byte(src[i]));
                i += 1;
            }
            continue;
        }
        if c == quote {
            return (String::from_utf8_lossy(&b).into_owned(), i + 1);
        }
        b.push(c);
        i += 1;
    }
    (String::from_utf8_lossy(&b).into_owned(), src.len())
}

/// Rust 原始字符串 `r"..."` / `r#"..."#` / `r##"..."##` … —— 跨行，零转义。
/// `i` 指向 `r`。成功返回（内容, 闭合之后下标, true）；非 raw 形式返回 (_, i, false)。
fn scan_raw(src: &[u8], i: usize) -> (String, usize, bool) {
    let mut j = i + 1;
    let mut hashes = 0usize;
    while j < src.len() && src[j] == b'#' {
        hashes += 1;
        j += 1;
    }
    if j >= src.len() || src[j] != b'"' {
        return (String::new(), i, false);
    }
    let content_start = j + 1;
    let mut k = content_start;
    while k < src.len() {
        if src[k] == b'"' {
            let mut h = 0;
            while h < hashes && k + 1 + h < src.len() && src[k + 1 + h] == b'#' {
                h += 1;
            }
            if h == hashes {
                let val = String::from_utf8_lossy(&src[content_start..k]).into_owned();
                return (val, k + 1 + hashes, true);
            }
        }
        k += 1;
    }
    (
        String::from_utf8_lossy(&src[content_start..]).into_owned(),
        src.len(),
        true,
    )
}

/// Rust 块注释 `/* ... */`（可嵌套）。`i` 指向第一个 `/`，返回（整段含定界符, 之后下标）。
fn scan_block_comment(src: &[u8], i: usize) -> (String, usize) {
    let mut j = i + 2;
    let mut depth = 1usize;
    while j < src.len() && depth > 0 {
        if j + 1 < src.len() && src[j] == b'/' && src[j + 1] == b'*' {
            depth += 1;
            j += 2;
            continue;
        }
        if j + 1 < src.len() && src[j] == b'*' && src[j + 1] == b'/' {
            depth -= 1;
            j += 2;
            continue;
        }
        j += 1;
    }
    (String::from_utf8_lossy(&src[i..j]).into_owned(), j)
}

/// 跨行字面量消费后同步 line/line_start（字面量内部换行不产生 Newline Token，但行号需前进）。
fn advance_line_count(
    src: &[u8],
    start: usize,
    end: usize,
    line: &mut i32,
    line_start: &mut usize,
) {
    for k in start..end {
        if src[k] == b'\n' {
            *line += 1;
            *line_start = k + 1;
        }
    }
}

fn is_token_delim(c: u8) -> bool {
    matches!(
        c,
        b' ' | b'\t'
            | b'\n'
            | b'\r'
            | b';'
            | b','
            | b')'
            | b'('
            | b'{'
            | b'}'
            | b'['
            | b']'
            | 0xC2 // ·（U+00B7）成员分隔符首字节
            | b'+'
            | b'-'
            | b'*'
            | b'%'
            | b'!'
            | b'='
            | b'<'
            | b'>'
            | b'&'
            | b'|'
            | b'^'
            | b':'
    )
}

fn is_abs_path_start(c: u8) -> bool {
    (b'a'..=b'z').contains(&c)
        || (b'A'..=b'Z').contains(&c)
        || (b'0'..=b'9').contains(&c)
        || c == b'_'
}

/// 将整个源扫描为平坦 Token 流，末尾附 EOF。
pub fn scan(src: &str) -> Vec<Token> {
    let src = src.as_bytes();
    let mut tokens: Vec<Token> = Vec::new();
    let mut i = 0usize;
    let mut line = 1i32;
    let mut line_start = 0usize;
    let mut prev_newline = true;

    let pos = |i: usize, line: i32, line_start: usize| Pos {
        line,
        col: (i - line_start) as i32 + 1,
    };

    while i < src.len() {
        let c = src[i];

        // 换行：折叠连续换行
        if c == b'\n' {
            if !prev_newline && !tokens.is_empty() {
                tokens.push(Token {
                    kind: Kind::Newline,
                    value: "\n".to_string(),
                    pos: pos(i, line, line_start),
                    quote: 0,
                });
                prev_newline = true;
            }
            i += 1;
            line += 1;
            line_start = i;
            continue;
        }
        if c == b'\r' {
            i += 1;
            continue;
        }
        // 空白（非换行）
        if c == b' ' || c == b'\t' {
            i += 1;
            continue;
        }
        // ';' — 显式语句分隔符
        if c == b';' {
            if !prev_newline && !tokens.is_empty() {
                tokens.push(Token {
                    kind: Kind::Newline,
                    value: ";".to_string(),
                    pos: pos(i, line, line_start),
                    quote: 0,
                });
                prev_newline = true;
            }
            i += 1;
            continue;
        }
        prev_newline = false;
        let p = pos(i, line, line_start);

        // Rust 原始字符串 r"..." / r#"..."# — 跨行，零转义
        if c == b'r' {
            let (val, next, ok) = scan_raw(src, i);
            if ok {
                tokens.push(Token {
                    kind: Kind::Literal,
                    value: val,
                    pos: p,
                    quote: b'r',
                });
                advance_line_count(src, i, next, &mut line, &mut line_start);
                i = next;
                continue;
            }
        }

        // 引号字符串 "..." — 跨行，转义
        if c == b'\'' || c == b'"' {
            let (val, next) = scan_quoted(src, i, c);
            let quote = if c == b'"' { b'"' } else { 0 };
            tokens.push(Token {
                kind: Kind::Literal,
                value: val,
                pos: p,
                quote,
            });
            advance_line_count(src, i, next, &mut line, &mut line_start);
            i = next;
            continue;
        }

        // 左箭头 <-
        if c == b'<' && i + 1 < src.len() && src[i + 1] == b'-' {
            tokens.push(Token {
                kind: Kind::Arrow,
                value: "<-".to_string(),
                pos: p,
                quote: 0,
            });
            i += 2;
            continue;
        }
        // 右箭头 ->
        if c == b'-' && i + 1 < src.len() && src[i + 1] == b'>' {
            tokens.push(Token {
                kind: Kind::Arrow,
                value: "->".to_string(),
                pos: p,
                quote: 0,
            });
            i += 2;
            continue;
        }

        // 双字符算子
        if i + 1 < src.len() {
            let two = &src[i..i + 2];
            if symbol::scanner_two_char_ops()
                .iter()
                .any(|op| op.as_bytes() == two)
            {
                tokens.push(Token {
                    kind: Kind::Ident,
                    value: String::from_utf8_lossy(two).into_owned(),
                    pos: p,
                    quote: 0,
                });
                i += 2;
                continue;
            }
        }

        // 赋值算子 =
        if c == b'=' {
            tokens.push(Token {
                kind: Kind::Arrow,
                value: "=".to_string(),
                pos: p,
                quote: 0,
            });
            i += 1;
            continue;
        }

        // '/' — // 行注释 或 /* */ 块注释 或 绝对路径字面量 或 除法算子
        if c == b'/' {
            if i + 1 < src.len() && src[i + 1] == b'/' {
                let start = i;
                while i < src.len() && src[i] != b'\n' {
                    i += 1;
                }
                tokens.push(Token {
                    kind: Kind::Comment,
                    value: String::from_utf8_lossy(&src[start..i]).into_owned(),
                    pos: p,
                    quote: 0,
                });
                continue;
            }
            if i + 1 < src.len() && src[i + 1] == b'*' {
                let (val, next) = scan_block_comment(src, i);
                tokens.push(Token {
                    kind: Kind::Comment,
                    value: val,
                    pos: p,
                    quote: 0,
                });
                advance_line_count(src, i, next, &mut line, &mut line_start);
                i = next;
                continue;
            }
            if i + 1 < src.len() && is_abs_path_start(src[i + 1]) {
                let start = i;
                i += 1;
                while i < src.len() {
                    // ·（U+00B7）与 .（释放给小数 key）是路径字符；· 后的坐标段 [0,1] 是成员链。
                    if src[i] == 0xC2 && i + 1 < src.len() && src[i + 1] == 0xB7 {
                        i += 2;
                        if i < src.len() && src[i] == b'[' {
                            while i < src.len() && src[i] != b']' {
                                i += 1;
                            }
                            if i < src.len() {
                                i += 1; // 跳过 ]
                            }
                        }
                        continue;
                    }
                    if src[i] == b'.' {
                        i += 1;
                        continue;
                    }
                    if is_token_delim(src[i]) {
                        break;
                    }
                    i += 1;
                }
                tokens.push(Token {
                    kind: Kind::Literal,
                    value: String::from_utf8_lossy(&src[start..i]).into_owned(),
                    pos: p,
                    quote: 0,
                });
            } else {
                tokens.push(Token {
                    kind: Kind::Ident,
                    value: "/".to_string(),
                    pos: p,
                    quote: 0,
                });
                i += 1;
            }
            continue;
        }

        // 成员分隔符 ·（U+00B7，2 字节）
        if c == 0xC2 && i + 1 < src.len() && src[i + 1] == 0xB7 {
            tokens.push(Token {
                kind: Kind::Dot,
                value: "·".to_string(),
                pos: p,
                quote: 0,
            });
            i += 2;
            prev_newline = false;
            continue;
        }

        // 单字符标点
        if let Some(k) = single_char_token(c) {
            tokens.push(Token {
                kind: k,
                value: (c as char).to_string(),
                pos: p,
                quote: 0,
            });
            i += 1;
            continue;
        }

        // 单字符符号算子
        if symbol::scanner_one_char_ops().contains(&c) {
            tokens.push(Token {
                kind: Kind::Ident,
                value: (c as char).to_string(),
                pos: p,
                quote: 0,
            });
            i += 1;
            continue;
        }

        // Unicode 符号算子
        if let Some(op) = symbol::unicode_ops()
            .into_iter()
            .find(|op| src[i..].starts_with(op.as_bytes()))
        {
            tokens.push(Token {
                kind: Kind::Ident,
                value: op.clone(),
                pos: p,
                quote: 0,
            });
            i += op.len();
            continue;
        }

        // 数字字面量
        if c.is_ascii_digit() {
            let start = i;
            while i < src.len() && src[i].is_ascii_digit() {
                i += 1;
            }
            if i < src.len() && src[i] == b'.' {
                i += 1;
                while i < src.len() && src[i].is_ascii_digit() {
                    i += 1;
                }
            }
            if i < src.len() && (src[i] == b'e' || src[i] == b'E') {
                i += 1;
                if i < src.len() && (src[i] == b'+' || src[i] == b'-') {
                    i += 1;
                }
                while i < src.len() && src[i].is_ascii_digit() {
                    i += 1;
                }
            }
            tokens.push(Token {
                kind: Kind::Literal,
                value: String::from_utf8_lossy(&src[start..i]).into_owned(),
                pos: p,
                quote: 0,
            });
            continue;
        }

        // 关键字 / 标识符
        let start = i;
        while i < src.len() && !is_token_delim(src[i]) {
            i += 1;
        }
        if i == start {
            i += 1;
            continue;
        }
        let word = String::from_utf8_lossy(&src[start..i]).into_owned();
        if let Some(k) = keyword(&word) {
            tokens.push(Token {
                kind: k,
                value: word,
                pos: p,
                quote: 0,
            });
        } else {
            tokens.push(Token {
                kind: Kind::Ident,
                value: word,
                pos: p,
                quote: 0,
            });
        }
    }

    tokens.push(Token::eof(pos(i, line, line_start)));
    tokens
}
